Makefile

```diff
@@ -141,66 +124,28 @@ tfgen: install_plugins upstream
 	$(WORKING_DIR)/bin/$(TFGEN) schema --out provider/cmd/$(PROVIDER)
 	(cd provider && VERSION=$(VERSION) go generate cmd/$(PROVIDER)/main.go)
 
-.PHONY: bin/pulumi-java-gen
-bin/pulumi-java-gen:
-# If $@ is already downloaded and then version doesn't match, we remove the old version.
-	@if [ -f $@ ]; then \
-		if [ $$($@ version) != $(JAVA_GEN_VERSION) ]; then \
-			echo "$@ version mismatch... redownloading"; \
-			rm $@; \
-		fi \
-	fi
-# We download the correct version when $@ is not present. That means either that
-# 1. $@ never existed, or
-# 2. $@ had the wrong version, and so was deleted.
-	@if ! [ -f $@ ]; then \
-		echo "Downloading $@ at version $(JAVA_GEN_VERSION)"; \
-		pulumictl download-binary -n pulumi-language-java -v $(JAVA_GEN_VERSION) -r pulumi/pulumi-java; \
-	fi
-
-init_upstream:
-	@# If the source is set, remove it. We do this since setting the source
-	@# when it is already set is an error.
-	@if [ ! -f "upstream/.git" ]; then \
-		echo "Initializing upstream submodule" && \
-		cd upstream && \
-		git submodule update --init && \
-		(git remote rm source || true) && \
-		git remote add source https://github.com/hashicorp/terraform-provider-aws.git && \
-		git fetch source; \
-	fi; \
-
-export_upstream_patches: init_upstream
-ifeq ($(shell cd upstream && git rev-parse --is-shallow-repository), false)
-	find upstream-patches -type f -delete
-	# Find the most recent tag before the current checkout - don't abbreviate the tag name
-	# Create patch files for each commit since the last tag
-	cd upstream && LAST_TAG=$$(git describe --abbrev=0 --tags) && \
-		git format-patch  -o ../upstream-patches --minimal --no-signature HEAD...$${LAST_TAG}
+upstream::
+# Use upstream.sh script to apply patches
+ifneq ("$(wildcard upstream)","")
+	@$(SHELL) ./scripts/upstream.sh "$@" apply
 endif
 
-upstream: init_upstream export_upstream_patches
+upstream::
```

Comment:
Instead of including this twice in the same Makefile, can we instead put a guarded [include](https://www.gnu.org/software/make/manual/html_node/Include.html) at the top and bottom of all our generated Makefiles, and supply this upstream definition there?

We could put this before we start defining targets:
```make
-include pre.mk
```

After we have defined all targets, we could put this:
```make
-include post.mk
```

In this case, `post.mk` would be:
```make
upstream::
	# Ensure tool is installed
	cd upstream-tools && yarn install --frozen-lockfile
	# Apply all automated changes
	cd upstream-tools && yarn --silent run apply
	# Check for any pending replacements
	cd upstream-tools && yarn --silent run check
```

`pre.mk` and `post.mk` would be checked in the provider repo, but never seen by ci-mgmt.

Reply:
Why do we want a pre.mk?